hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/procedures/ShowMetadataTableHistoryProcedure.scala

```diff
@@ -0,0 +1,293 @@
+                                              action: String,
+                                              requestedTime: String,
+                                              completionTime: String,
+                                              modificationTimeMs: Long
+                                             ) {
+    def getModificationTime: Long = modificationTimeMs
+  }
+
+  private def buildInstantInfoFromTimeline(metaClient: HoodieTableMetaClient): Map[String, Map[HoodieInstant.State, HoodieInstantWithModTime]] = {
+    try {
+      val storage = metaClient.getStorage
+      val timelinePath = metaClient.getTimelinePath
+      val instantFileNameParser = metaClient.getInstantFileNameParser
+      val instantGenerator = metaClient.getInstantGenerator
+
+      val instantMap = scala.collection.mutable.Map[String, scala.collection.mutable.Map[HoodieInstant.State, HoodieInstantWithModTime]]()
+
+      val fileStream = HoodieTableMetaClient.scanFiles(storage, timelinePath, path => {
+        val extension = instantFileNameParser.getTimelineFileExtension(path.getName)
+        metaClient.getActiveTimeline.getValidExtensionsInActiveTimeline.contains(extension)
+      })
+
+      fileStream.forEach { storagePathInfo =>
+        try {
+          val instant = instantGenerator.createNewInstant(storagePathInfo)
+          val instantWithModTime = HoodieInstantWithModTime(
+            instant.getState,
+            instant.getAction,
+            instant.requestedTime(),
+            instant.getCompletionTime,
+            storagePathInfo.getModificationTime
+          )
+
+          instantMap.getOrElseUpdate(instant.requestedTime(), scala.collection.mutable.Map[HoodieInstant.State, HoodieInstantWithModTime]())
+            .put(instant.getState, instantWithModTime)
+        } catch {
+          case _: Exception => // Skip invalid files
+        }
+      }
+      instantMap.map { case (timestamp, stateMap) =>
+        timestamp -> stateMap.toMap
+      }.toMap
+    } catch {
+      case _: Exception => Map.empty[String, Map[HoodieInstant.State, HoodieInstantWithModTime]]
+    }
+  }
+
+  private def getFormattedDateForState(instantTimestamp: String, state: HoodieInstant.State,
+                                       instantInfoMap: Map[String, Map[HoodieInstant.State, HoodieInstantWithModTime]]): String = {
+    val stateMap = instantInfoMap.get(instantTimestamp)
+    if (stateMap.isDefined) {
+      val stateInfo = stateMap.get.get(state)
+      if (stateInfo.isDefined) {
+        val modificationTime = stateInfo.get.getModificationTime
+        if (modificationTime > 0) {
+          val date = new java.util.Date(modificationTime)
+          val formatter = new java.text.SimpleDateFormat("MM-dd HH:mm:ss")
+          formatter.format(date)
+        } else {
+          instantTimestamp
+        }
+      } else {
+        null
+      }
+    } else {
+      null
+    }
+  }
```

Comment:
We can simplify this to
```suggestion
    val formatter = new java.text.SimpleDateFormat("MM-dd HH:mm:ss")
    instantInfoMap.get(instantTimestamp)
      .flatMap(_.get(state))
      .map(_.modificationTimeMs)
      .filter(_ > 0)
      .map(time => formatter.format(new java.util.Date(time)))
      .getOrElse(instantTimestamp)

```